mitxgraders/comparers/comparers.py

```diff
@@ -112,30 +113,108 @@ class EqualityComparer(Comparer):
     True
-    >>> FormulaGrader.reset_default_comparer()
+    >>> MatrixGrader.reset_default_comparer()
+
+    FormulaGrader and MatrixGrader both have set_default_comparer() and
+    reset_default_comparer() methods.
 
     """
     schema_config = Schema({
         Required('transform', default=None): Nullable(is_callable)
     })
 
-    def __call__(self, comparer_params_eval, student_eval, utils):
-        expected_input = comparer_params_eval[0]
-
+    @staticmethod
+    def validate(expected_eval, student_eval, utils):
         if hasattr(utils, 'validate_shape'):
             # in numpy, scalars have empty tuples as their shapes
-            shape = tuple() if isinstance(expected_input, Number) else expected_input.shape
+            shape = tuple() if isinstance(expected_eval, Number) else expected_eval.shape
             utils.validate_shape(student_eval, shape)
 
-        # If provided, apply the transform function
-        if self.config['transform'] is not None:
-            expected_input = self.config['transform'](expected_input)
-            student_eval = self.config['transform'](student_eval)
+    def transform(self, evaluation):
+        f = self.config['transform']
+        return evaluation if f is None else f(evaluation)
+
 
-        return utils.within_tolerance(expected_input, student_eval)
+    def __call__(self, comparer_params_eval, student_eval, utils):
+        expected_eval = comparer_params_eval[0]
+        self.validate(expected_eval, student_eval, utils)
+
+        expected_eval = self.transform(expected_eval)
+        student_eval = self.transform(student_eval)
+
+        return utils.within_tolerance(expected_eval, student_eval)
 
 equality_comparer = EqualityComparer()
 
+class MatrixEntryComparer(EqualityComparer):
+    """
+    Default comparer for MatrixGrader. Compares student and instructor matirx
+    evaluations entry-by-entry for equality.
+
+    Configuration
+    =============
+
+        transform (None | function): same as EqualityComparer, defaults to None
+        entry_partial_credit ('proportional'|number): Determines how partial credit
+            is awarded. If set to 'proportional', then credit is proportional to
+            the number of correct matrix entries. If a numeric value is provided,
+            this flat rate of partial credit is provided as long as some but
+            not all entries are correct.
+
+            Default is the numeric value 0 (no partial credit).
+        entry_partial_msg (str): A text string message shown when partial credit
+            is awarded. The string may optionally contain the formatting key {error_indices},
+            which will be replaced with the indices of the incorrect matrix entries.
+    """
+
+    default_msg = "Matrix entries at indices {error_indices} are incorrect."
+    schema_config = EqualityComparer.schema_config.extend({
+        Required('entry_partial_credit', default=0): Any(All(Number, Range(0, 1)), 'proportional'),
```

Comment:
Huh, I tried doing this somewhere else, but I used `All(float, Range(0, 1))`. It wouldn't accept 0 or 1, so I had to add them in separately in the `Any`. Guess I should've used `Number`?

Reply:
Yeah, I saw where you did that. Both seem fine. Maybe it's worth making a validator `numeric_range(x, y)` just for this case.

I was a little suprised that just `Range(0, 1)` doesn't work, but it throws `TypeError`s instead of `ValueError`s. I guess it's still good to declare Number or float or int, though, since things besides numbers might be comparable to 0 and 1 in Python.